### ISSUE TITLE
OSX support

### ISSUE THREAD

**Issue by esclapez** [dshawul/solver#6]
I know there is no advertised OSX support for NebulaSEM, but I can't seem to get a successful run on MacOS.
The requirements are pretty basic and I have tested other softwares with similar requirements without issues, so I am wondering if you would have any idea.

A bit more background:
- Apple MacBook Pro with MacOS Sonoma and a Silicon M3 chip.
- Clang (15.0.0) or brew installed GNU (13.2.0) compilers
- Open-MPI 5.0.2 or MPICH 4.2.1
Metis installed locally with the same combination of Compiler+MPI as what is used to build NebulaSEM. With all the combinations, I am able to compile the apps, but always run into hangs when running `convection` (the piso cavity case seems to run without issues).

For instance, when I run `./test.sh -n 2 -c examples/atmo/advection-leveque/bubble`, the code hangs at:
```
...
--------------------------------------------
Reading T2
Reading U2
--------------------------------------------
2247 [0] Starting iterations.
2248 [0] Time 0.125625
2254 [0] SYMM-DIAG-DIAG-JAC:Iterations 1 Initial Residual 0.00000e+00 Final Residual 0.00000e+00
2254 [0] Scalar loss: -1404947.06142 Volume loss: 0.525714285714
2254 [0] Time 0.126250
2260 [0] SYMM-DIAG-DIAG-JAC:Iterations 1 Initial Residual 0.00000e+00 Final Residual 0.00000e+00
2261 [0] Scalar loss: -1404906.22076 Volume loss: 0.525714285714
2456 [0] Time 0.146250
2462 [0] SYMM-DIAG-DIAG-JAC:Iterations 1 Initial Residual 0.00000e+00 Final Residual 0.00000e+00
2462 [0] Scalar loss: -1402350.15106 Volume loss: 0.525714285714
2657 [0] Time 0.166250
2663 [0] SYMM-DIAG-DIAG-JAC:Iterations 1 Initial Residual 0.00000e+00 Final Residual 0.00000e+00
2663 [0] Scalar loss: -1395880.33326 Volume loss: 0.525714285714
2858 [0] Time 0.186250
2864 [0] SYMM-DIAG-DIAG-JAC:Iterations 1 Initial Residual 0.00000e+00 Final Residual 0.00000e+00
2864 [0] Scalar loss: -1383093.35153 Volume loss: 0.525714285714
Merging fields at step 3
2882 [0] 	790 vertices	1441 facets	349 cells
Refining grid at step 3
2973 [0] 	790 vertices	1441 facets	349 cells
Refining 89 Coarsening 116
```

I have tried a `Debug` build (using GCC+OpenMPI) without much success.

Do you have any intuition of where the issue could come from ? The Scalar loss appears large but I might not be familiar with what it represents. I'd be happy to dig a bit further.

Thanks in advance.

**Comment by danielabdi-noaa** [comment]
Hi @esclapez 

Thank you testing it on Mac! I have never actually tried to compile and run it on Mac before today.
It looks like the examples work fine if you use 1 mpi rank but not with more.
I have run successfully above test case you run modifying `-n 1` and it run to completion and other test cases as well.
With 2 mpi ranks though, I get the same error as yours so need to debug that.
Can you confirm you are able to run the test cases with 1 mpi rank.

regards,

**Comment by esclapez** [comment]
Yes, indeed ! It runs fine with a single rank.

**Comment by danielabdi-noaa** [comment]
Thank you for confirming! I will try to test it with a different MPI implementation say MPICH instead of OpenMPI to see if that is the issue. Otherwise, will have to look into src/mp code to see what should be different! I will let you know once it is fixed but may take time to figure out.

Thanks again.

**Comment by danielabdi-noaa** [comment]
@esclapez It turns out Mac was not the issue, but that my MPI implementation had several bugs mostly because the code is old and I did not have CI/CD for the MPI code. The bugs are mostly related to AMR + MPI combo. AMR was my recent focus and forgot to test it with MPI. The current implementation is not robust and I plan to replace it with a mature library  such as p4est/AMReX eventually. Regarding comparison with existing software, I have just submitted a paper comparing cell-based AMR with this software and level-based AMR with AMReX specifically for numerical weather prediction. 

The arxiv link of the paper is here: https://arxiv.org/pdf/2404.16648

Anyway, I have fixed most of the bugs now and this test case should run with 2 or more MPI ranks. Other test cases should also work too but I will continue working on it.

Thanks again for spotting this issue!
